### PR TITLE
Basic map selection

### LINKED ISSUE #11: CLI command to generate PMTiles from GerryDB view(s)
**Todo:**

- [ ] Write and test command
  - [x] locally
  - [ ] on prod
  - [x] Command should create copy as fgb first
- [x] Build table to store tileset paths and names
- [x] Build endpoint to fetch tileset paths (or `GET /api/document/{document_id}` should return path)
- [x] Create single-layer PMTiles for each received GerryDB view
- [x] Update bucket CORS

### LINKED ISSUE #12: Support map selection from available GerryDB views
For the midway user testing, we want to be able to be demonstrate perf for maps of different sizes. Support switching between received GerryDB views.

Nice to haves
- [ ] Zooming to the new selected map

Eventually this will be replaced by something similar to the [existing map selection flow](https://districtr.org/) in Districtr.

### PR DESCRIPTION
**Pull request by raphaellaude**
## Description

- Fixes #11 
- Fixes #12 
- Todo:
  - [x] Local runs (will only work for @raphaellaude locally due to R2 permissions)
  - [ ] Install tippecanoe in prod image (really though? should we use a worker or something? Don't want to block up the main thread...)

## Reviewers

- Primary: @mariogiampieri @bailliekova 
- Secondary:

## Checklist

- [ ] Added/Updated related documentation (if applicable).
- [x] Added/Updated related unit tests (if applicable).

### CODE REVIEW

**Comment by raphaellaude** [comment]
Successful local runs:
```
python cli.py create-gerrydb-tileset -g /Users/raphaellaude/Downloads/8f8f71971337445480a802cdfd8a951a.gpkg --layer ks_demo_view_census_blocks
['http://localhost:3000', 'http://localhost:5173', 'http://127.0.0.1:3000']
INFO:__main__:Creating GerryDB tileset...
INFO:__main__:URL: ParseResult(scheme='', netloc='', path='/Users/raphaellaude/Downloads/8f8f71971337445480a802cdfd8a951a.gpkg', params='', query='', fragment='')
INFO:__main__:Creating flatgeobuf...
INFO:__main__:File already exists. Skipping creation.
INFO:__main__:Creating tileset...
INFO:__main__:File already exists. Skipping creation.
INFO:__main__:Uploading to R2
INFO:__main__:GerryDB tileset uploaded successfully
INFO:__main__:Updating GerryDBTiles
2024-08-01 00:36:51,128 INFO sqlalchemy.engine.Engine select pg_catalog.version()
INFO:sqlalchemy.engine.Engine:select pg_catalog.version()
2024-08-01 00:36:51,128 INFO sqlalchemy.engine.Engine [raw sql] {}
INFO:sqlalchemy.engine.Engine:[raw sql] {}
2024-08-01 00:36:51,130 INFO sqlalchemy.engine.Engine select current_schema()
INFO:sqlalchemy.engine.Engine:select current_schema()
2024-08-01 00:36:51,130 INFO sqlalchemy.engine.Engine [raw sql] {}
INFO:sqlalchemy.engine.Engine:[raw sql] {}
2024-08-01 00:36:51,131 INFO sqlalchemy.engine.Engine show standard_conforming_strings
INFO:sqlalchemy.engine.Engine:show standard_conforming_strings
2024-08-01 00:36:51,131 INFO sqlalchemy.engine.Engine [raw sql] {}
INFO:sqlalchemy.engine.Engine:[raw sql] {}
2024-08-01 00:36:51,134 INFO sqlalchemy.engine.Engine BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
2024-08-01 00:36:51,135 INFO sqlalchemy.engine.Engine
        INSERT INTO gerrydbtiles (uuid, name, s3_path, updated_at)
        VALUES (gen_random_uuid(), %(name)s, %(s3_path)s, now())
        ON CONFLICT (name)
        DO UPDATE SET
            updated_at = now()
        RETURNING uuid

INFO:sqlalchemy.engine.Engine:
        INSERT INTO gerrydbtiles (uuid, name, s3_path, updated_at)
        VALUES (gen_random_uuid(), %(name)s, %(s3_path)s, now())
        ON CONFLICT (name)
        DO UPDATE SET
            updated_at = now()
        RETURNING uuid

2024-08-01 00:36:51,135 INFO sqlalchemy.engine.Engine [generated in 0.00016s] {'name': 'ks_demo_view_census_blocks', 's3_path': 'tilesets/ks_demo_view_census_blocks.pmtiles'}
INFO:sqlalchemy.engine.Engine:[generated in 0.00016s] {'name': 'ks_demo_view_census_blocks', 's3_path': 'tilesets/ks_demo_view_census_blocks.pmtiles'}
2024-08-01 00:36:51,139 INFO sqlalchemy.engine.Engine COMMIT
INFO:sqlalchemy.engine.Engine:COMMIT
INFO:__main__:GerryDB tileset upserted successfully:
(UUID('43df220a-555f-4044-b82c-6eed838ca4a4'),)
aws s3 ls s3://districtr-v2-dev/ --endpoint-url=https://de4ecd9d308a46631d2b677af1d480a0.r2.cloudflarestorage.com --profile cloudflare
                           PRE tilesets/
2024-07-18 22:11:35   27545600 488c6946a0d5487eaad3a801df8862db.gpkg
2024-07-18 22:12:03   40988672 89bb5423a5784d35a57d290f63b687a0.gpkg
2024-07-18 22:17:08  213524480 8f8f71971337445480a802cdfd8a951a.gpkg
2024-07-18 22:10:50    2928640 c22cea3c83e14fd79be9a69c87dd089c.gpkg
2024-07-18 22:11:14   16130048 c6c23a64a3234171853c9897095f001b.gpkg
aws s3 ls s3://districtr-v2-dev/tilesets --endpoint-url=https://de4ecd9d308a46631d2b677af1d480a0.r2.cloudflarestorage.com --profile cloudflare
                           PRE tilesets/
aws s3 ls s3://districtr-v2-dev/tilesets/ --endpoint-url=https://de4ecd9d308a46631d2b677af1d480a0.r2.cloudflarestorage.com --profile cloudflare
2024-08-01 00:36:50          0
2024-08-01 00:36:51       1735 ks_demo_view_census_blocks.pmtiles
python cli.py create-gerrydb-tileset -g s3://districtr-v2-dev/8f8f71971337445480a802cdfd8a951a.gpkg --layer ks_demo_view_census_blocks
['http://localhost:3000', 'http://localhost:5173', 'http://127.0.0.1:3000']
INFO:__main__:Creating GerryDB tileset...
INFO:__main__:URL: ParseResult(scheme='s3', netloc='districtr-v2-dev', path='/8f8f71971337445480a802cdfd8a951a.gpkg', params='', query='', fragment='')
INFO:__main__:File name: 8f8f71971337445480a802cdfd8a951a.gpkg
INFO:__main__:Importing GerryDB view. Got response:
{'ResponseMetadata': {'HTTPStatusCode': 200, 'HTTPHeaders': {'date': 'Thu, 01 Aug 2024 04:38:23 GMT', 'content-type': 'application/x-sqlite3', 'content-length': '213524480', 'connection': 'keep-alive', 'accept-ranges': 'bytes', 'etag': '"0985a83dbfa4c188e0d96e601944ab68"', 'last-modified': 'Fri, 19 Jul 2024 02:17:08 GMT', 'x-amz-storage-class': 'STANDARD_IA', 'vary': 'Accept-Encoding', 'server': 'cloudflare', 'cf-ray': '8ac3236debd16a5b-EWR'}, 'RetryAttempts': 0}, 'AcceptRanges': 'bytes', 'LastModified': datetime.datetime(2024, 7, 19, 2, 17, 8, tzinfo=tzutc()), 'ContentLength': 213524480, 'ETag': '"0985a83dbfa4c188e0d96e601944ab68"', 'ContentType': 'application/x-sqlite3', 'Metadata': {}, 'StorageClass': 'STANDARD_IA'}
INFO:__main__:Downloading file...
INFO:__main__:Creating flatgeobuf...
INFO:__main__:File already exists. Skipping creation.
INFO:__main__:Creating tileset...
INFO:__main__:File already exists. Skipping creation.
INFO:__main__:Uploading to R2
INFO:__main__:GerryDB tileset uploaded successfully
INFO:__main__:Updating GerryDBTiles
2024-08-01 00:38:31,858 INFO sqlalchemy.engine.Engine select pg_catalog.version()
INFO:sqlalchemy.engine.Engine:select pg_catalog.version()
2024-08-01 00:38:31,858 INFO sqlalchemy.engine.Engine [raw sql] {}
INFO:sqlalchemy.engine.Engine:[raw sql] {}
2024-08-01 00:38:31,860 INFO sqlalchemy.engine.Engine select current_schema()
INFO:sqlalchemy.engine.Engine:select current_schema()
2024-08-01 00:38:31,860 INFO sqlalchemy.engine.Engine [raw sql] {}
INFO:sqlalchemy.engine.Engine:[raw sql] {}
2024-08-01 00:38:31,861 INFO sqlalchemy.engine.Engine show standard_conforming_strings
INFO:sqlalchemy.engine.Engine:show standard_conforming_strings
2024-08-01 00:38:31,861 INFO sqlalchemy.engine.Engine [raw sql] {}
INFO:sqlalchemy.engine.Engine:[raw sql] {}
2024-08-01 00:38:31,864 INFO sqlalchemy.engine.Engine BEGIN (implicit)
INFO:sqlalchemy.engine.Engine:BEGIN (implicit)
2024-08-01 00:38:31,865 INFO sqlalchemy.engine.Engine
        INSERT INTO gerrydbtiles (uuid, name, s3_path, updated_at)
        VALUES (gen_random_uuid(), %(name)s, %(s3_path)s, now())
        ON CONFLICT (name)
        DO UPDATE SET
            updated_at = now()
        RETURNING uuid

INFO:sqlalchemy.engine.Engine:
        INSERT INTO gerrydbtiles (uuid, name, s3_path, updated_at)
        VALUES (gen_random_uuid(), %(name)s, %(s3_path)s, now())
        ON CONFLICT (name)
        DO UPDATE SET
            updated_at = now()
        RETURNING uuid

2024-08-01 00:38:31,865 INFO sqlalchemy.engine.Engine [generated in 0.00016s] {'name': 'ks_demo_view_census_blocks', 's3_path': 'tilesets/ks_demo_view_census_blocks.pmtiles'}
INFO:sqlalchemy.engine.Engine:[generated in 0.00016s] {'name': 'ks_demo_view_census_blocks', 's3_path': 'tilesets/ks_demo_view_census_blocks.pmtiles'}
2024-08-01 00:38:31,867 INFO sqlalchemy.engine.Engine COMMIT
INFO:sqlalchemy.engine.Engine:COMMIT
INFO:__main__:GerryDB tileset upserted successfully:
(UUID('43df220a-555f-4044-b82c-6eed838ca4a4'),)
aws s3 ls s3://districtr-v2-dev/tilesets/ --endpoint-url=https://de4ecd9d308a46631d2b677af1d480a0.r2.cloudflarestorage.com --profile cloudflare
2024-08-01 00:38:31          0
2024-08-01 00:38:31       1735 ks_demo_view_census_blocks.pmtiles
```

**Comment by raphaellaude** [comment]
I've loaded the following views to prod:
```
districtr_v2_api=# \dt gerrydb.*
                        List of relations
 Schema  |            Name            | Type  |      Owner
---------+----------------------------+-------+------------------
 gerrydb | de_demo_view_census_blocks | table | districtr_v2_api
 gerrydb | de_demo_view_census_vtd    | table | districtr_v2_api
 gerrydb | ks_demo_view_census_blocks | table | districtr_v2_api
 gerrydb | ks_demo_view_census_vtd    | table | districtr_v2_api
 gerrydb | pa_demo_view_census_vtd    | table | districtr_v2_api
(5 rows)
```

And the following tilesets to the prod bucket
```
aws s3 ls s3://districtr-v2-dev/tilesets/ --endpoint-url=https://de4ecd9d308a46631d2b677af1d480a0.r2.cloudflarestorage.com --profile cloudflare
2024-08-02 17:54:22          0
2024-08-02 17:54:23       1874 de_demo_view_census_blocks.pmtiles
2024-08-02 17:54:01       2182 de_demo_view_census_vtd.pmtiles
2024-08-02 17:54:08       1735 ks_demo_view_census_blocks.pmtiles
2024-08-02 17:53:56       2029 ks_demo_view_census_vtd.pmtiles
2024-08-02 17:54:15       1929 pa_demo_view_census_vtd.pmtiles
```

**Comment by raphaellaude** [comment]
@mariogiampieri would you mind taking a look at this since you've done some NextJS stuff with PMTiles? I'm having the hardest time getting the tiles to show up. Seeing this error, from w/i pmtiles, when I add the source:
```
index.mjs:746 Uncaught (in promise) TypeError: callback is not a function
    at eval (index.mjs:746:11)
```
Haven't had this issue with Vite.

File: app/.env.dev [hunk 1]
```diff
@@ -1 +1,2 @@
 NEXT_PUBLIC_API_URL=http://localhost:8000
+NEXT_PUBLIC_S3_BUCKET_URL=https://de4ecd9d308a46631d2b677af1d480a0.r2.cloudflarestorage.com/districtr-v2-dev
```
Comment (bailliekova) [inline app/.env.dev:2]: (O) do we need to obscure this at all?

File: backend/app/alembic/env.py [hunk 2]
```diff
@@ -43,7 +44,7 @@ def get_url():
 
 
 def include_object(object, name, type_, reflected, compare_to):
-    if name == "spatial_ref_sys":
+    if name and (name == "spatial_ref_sys" or re.match(r"assignments_.+", name)):
```
Comment (bailliekova) [inline backend/app/alembic/env.py:47]: (pp) add a comment here

File: backend/app/main.py [hunk 3]
```diff
@@ -146,3 +149,23 @@ async def get_total_population(
...
+    session: Session = Depends(get_session),
+    offset: int = 0,
+    limit: int = Query(default=100, le=100),
+):
+    gerrydb_views = session.exec(
+        select(
+            GerryDBTable.name.label("table_name"),
+            GerryDBTiles.s3_path.label("tiles_s3_path"),
+        )
+        .join(GerryDBTiles, GerryDBTable.uuid == GerryDBTiles.table_uuid, isouter=True)
```
Comment (bailliekova) [inline backend/app/main.py:166]: (o) I think we had talked about upstream validation and inner joins? or is there a different case where we might have unmatched rows?

File: backend/app/main.py [hunk 4]
```diff
@@ -146,3 +149,23 @@ async def get_total_population(
...
+        .order_by(GerryDBTable.created_at.asc())
```
Comment (bailliekova) [inline backend/app/main.py:167]: (o) do we need the order by?  IMO lose it if not?

File: backend/app/models.py [hunk 5]
```diff
@@ -35,6 +44,20 @@ class GerryDBTable(TimeStampMixin, SQLModel, table=True):
     name: str = Field(nullable=False, unique=True)
 
 
+# Since `table_uuid` foreign key is unique, is there a good reason for tiles to be a separate table?
+class GerryDBTiles(TimeStampMixin, SQLModel, table=True):
+    uuid: str = Field(sa_column=Column(UUIDType, unique=True, primary_key=True))
+    table_uuid: str = Field(
+        sa_column=Column(UUIDType, ForeignKey(GerryDBTable.uuid), unique=True)
+    )
+    s3_path: str = Field(nullable=False)
```
Comment (bailliekova) [inline backend/app/models.py:53]: (q) what's the difference between this s3 path and the tiles_s3 path in the GerryDBViewPublic class?

File: backend/cli.py [hunk 6]
```diff
@@ -19,45 +18,48 @@ def cli():
...
+
+    return path
+
+
 @cli.command("import-gerrydb-view")
 @click.option("--layer", "-n", help="layer of the view", required=True)
 @click.option("--gpkg", "-g", help="Path or URL to GeoPackage file", required=True)
 @click.option("--replace", "-f", help="Replace the file if it exists", is_flag=True)
 @click.option("--rm", "-r", help="Delete file after loading to postgres", is_flag=True)
 def import_gerrydb_view(layer: str, gpkg: str, replace: bool, rm: bool):
```
Comment (bailliekova) [inline backend/cli.py:53]: (pp) can we add a comment here to clarify that import here means downloading to the volume but not doing any processing on the geopackage?

File: backend/cli.py [hunk 7]
```diff
@@ -119,5 +118,141 @@ def import_gerrydb_view(layer: str, gpkg: str, replace: bool, rm: bool):
...
+            gen_random_uuid(),
+            (SELECT uuid FROM gerrydbtable WHERE name = :name),
+            :s3_path,
+            now()
+        )
+        ON CONFLICT (table_uuid)
+        DO UPDATE SET
+            updated_at = now(),
+            s3_path = :s3_path
+        RETURNING uuid
```
Comment (bailliekova) [inline backend/cli.py:236]: 🎉 i love postgres's RETURNING keyword

File: backend/scripts/load_gerrydb_view.sh [hunk 8]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
Comment (bailliekova) [inline backend/scripts/load_gerrydb_view.sh:1]: I am very stupid, where is this invoked?

File: backend/scripts/print_gerrydb_layer_info.sh [hunk 9]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
```
Comment (bailliekova) [inline backend/scripts/print_gerrydb_layer_info.sh:1]: Same here, where is this invoked?

File: app/.env.dev (hunk 1 above)
Comment (raphaellaude) [reply app/.env.dev:2]: No – we'll be fetching resources from here so it'll be public regardless, even if we try to obscure it in the FE code w/ a secret. CORS should prevent abuse of the tilesets (once the bucket policy is set to be more stringent, i.e. only accept requests from the FE domain)

File: backend/app/main.py (hunk 3 above)
Comment (raphaellaude) [reply backend/app/main.py:166]: I was wondering about this and left it open for now. I was thinking maybe there's a case where we want to generate some metrics on a GerryDB view that isn't itself associated with a tileset?

File: backend/app/main.py (hunk 4 above)
Comment (raphaellaude) [reply backend/app/main.py:167]: Just made testing easier lol I can get rid of it ;)

File: backend/scripts/load_gerrydb_view.sh (hunk 8 above)
Comment (raphaellaude) [reply backend/scripts/load_gerrydb_view.sh:1]: Just locally for now. We should ask if there's a need for batch imports like I'm doing here on prod. For the time being I rather not add a command to do this, requiring admins to load layers 1 by 1.

File: backend/scripts/print_gerrydb_layer_info.sh (hunk 9 above)
Comment (raphaellaude) [reply backend/scripts/print_gerrydb_layer_info.sh:1]: idem above

**Comment by raphaellaude** [comment]
Works!
https://github.com/user-attachments/assets/aa55b6f6-d3ac-47d1-aa05-967ccbca65af

File: backend/app/models.py (hunk 5 above)
Comment (bailliekova) [reply backend/app/models.py:53]: IMO shouldn't store it twice to make consistency easier, and if we're finding ourselves always having to join both tables at the same granularity, makes sense to make it one from the get-go.

**Review by mariogiampieri — DISMISSED** [review]
lgtm

File: app/src/app/components/sidebar/Sidebar.tsx [hunk 10]
```diff
@@ -1,19 +1,18 @@
...
 import { ColorPicker } from "./ColorPicker";
 import { ResetMapButton } from "./ResetMapButton";
+import { GerryDBViewSelector } from "./GerryDBViewSelector";
 
 export default function SidebarComponent() {
   return (
-    <div className="h-full w-sidebar m-3 pt-3 py-1.5">
-      <div className="flex flex-col py-1.5">
+    <Box m="3">
+      <Flex direction="column" gap="3">
```
Comment (mariogiampieri) [inline app/src/app/components/sidebar/Sidebar.tsx:10]: much cleaner

File: app/src/app/constants/layers.ts [hunk 11]
```diff
@@ -25,58 +27,74 @@ ZONE_ASSIGNMENT_STYLE_DYNAMIC.push("#cecece");
...
+    },
+  };
+}
 
-  map.current?.addLayer(BLOCKS_HOVER_LAYER);
+const addBlockLayers = (
+  map: MutableRefObject<Map | null>,
+  gerryDBView: gerryDBView,
+) => {
+  const blockSource = getBlocksSource(gerryDBView.tiles_s3_path);
```
Comment (mariogiampieri) [inline app/src/app/constants/layers.ts:82]: i like this implementation, very cool

**Review by mariogiampieri — APPROVED** [review]
lgtm- well done!

File: backend/app/main.py (hunk 3 above)
Comment (raphaellaude) [reply backend/app/main.py]: got rid of join because I merged the tables

File: backend/app/alembic/env.py (hunk 2 above)
Comment (raphaellaude) [reply backend/app/alembic/env.py:47]: oop forgot to do this. leaving unresolved so I can get back to it

File: backend/app/models.py (hunk 5 above)
Comment (raphaellaude) [reply backend/app/models.py]: combined these two tables!

**Pull request merged**